SEED-platform/seed
# Geocoding fields

This PR includes migrations.

#### Any background context you want to provide?
There are cases where users don't store address information in the provided DB address columns such as `address_line_1`, `city`, `postal_code`, etc.

In those cases, geocoding wouldn't be possible as geocoding was "hardcoded" to use the provided DB addresses.

#### What's this PR do?
Geocoding columns are now customizable. Users can toggle columns to be a part of geocoding as well as choose the order in which the corresponding values are concatenated to construct a full address.

Since there's now no way to pre-filter records that will provide ambiguous geocoding results, say if an address like "101 Market Street, California" provides multiple valid geocoding results, any time multiple geocoding results are provided for an address, the record is not geocoded and `Geocoding Confidence` is populated with "Low - check address (Ambiguous)"

Along with this, text/warnings/advice for geocoding has been added or updated in both the "Geocode Selected" action modal on the inventory list pages and the column mapping page during the import file flow.

#### How should this be manually tested?
With a valid MapQuest API key, update geocoding columns via the column settings page, then import files while using these columns to hold address fields. **Note**: For an organization, when no columns are used for geocoding an inventory type, no geocoding is attempted at all for that org's inventory type. When the org has enough geocoding columns specified for an inventory type, each record must still have at least 1 column with non-empty string or number values.

**Follow these 'with and without API key' steps to view updated flows along with text and advice.**

Without a MapQuest API key attached to an org, 
1. Visit the column mapping page within the import file flow and see that no text is provided regarding geocoding
2. Visit the inventory list pages and see that geocoding isn't possible when you select records and try to "Geocode Selected".

With a MapQuest API key, 
1. Visit the column settings page for an org; view the geocoding advice at the top of the page; and alter the geocoding columns (use DB columns and extra_data columns).
2. Visit the column mapping page within the import file flow, and see the columns you've selected for geocoding as well as text/advice about geocoding.
3. Revisit the column settings page, and remove geocoding columns for any inventory type so that that inventory type has no geocoding columns.
4. Revisit the column mapping page and see the columns are no longer listed for the inventory type that has no geocoding columns. (If you deactivated them from both properties and tax lots, then all the geocoding-related text should be removed.)
5. Visit the inventory list page for the inventory type with no geocoding columns and try to "Geocode Selected", a warning should appear that geocoding isn't possible since there aren't enough columns.

#### What are the relevant tickets?
#1963 

#### Screenshots (if appropriate)

**Note**:  In addition to the above changes, I used this PR to...
1. refresh the VCR cassettes for the geocoding tests that actually hit the MapQuest API. See that several .yaml files were updated.
2. create lokalised translations for geocoding-related text that was previously untranslated.
3. addressed relevant "Geocode selected" (from inventory list pages) UI/UX improvements
